File: 1 - LinkedList/C++/4-CircularDoubleLinkedList/circularDoubleLinkedList.cpp
#include <iostream>
#include <string>

using namespace std;

// Node class representing each element in the circular doubly linked list
class Node {
public:
    string name;  // Item name
    int price;    // Item price
    int available; // Item availability
    Node* prev;   // Pointer to the previous node in the list
    Node* next;   // Pointer to the next node in the list

    // Constructor to initialize a new node
    Node(const string& name, int price, int available)
        : name(name), price(price), available(available), prev(nullptr), next(nullptr) {}
};

// LinkedList class to manage the circular doubly linked list
class LinkedList {
private:
    Node* head; // Pointer to the head of the list
    Node* tail; // Pointer to the tail of the list
    int size;   // Size of the list

public:
    // Constructor to initialize an empty list
    LinkedList() : head(nullptr), tail(nullptr), size(0) {}

    // Destructor to clean up the list
    ~LinkedList() {
        while (size > 0) {
            removeFirstNode();
        }
    }

    // Function to create a circular doubly linked list with the first node
    void createCircularDL(const string& name, int price, int available) {
        Node* newNode = new Node(name, price, available);
        newNode->next = newNode;
        newNode->prev = newNode;
        head = newNode;
        tail = newNode;
        size = 1;
    }

    // Function to add a node at the beginning of the list
    void addAtFirstNode(const string& name, int price, int available) {
        if (size == 0) {
            createCircularDL(name, price, available);
            return;
        }
        Node* newNode = new Node(name, price, available);
        newNode->next = head;
        newNode->prev = tail;
        head->prev = newNode;
        tail->next = newNode;
        head = newNode;
        size++;
    }

    // Function to add a node at the nth position in the list
    void addAtNthNode(const string& name, int price, int available, int pos) {
        if (pos < 1 || pos > size + 1) {
            cout << "Invalid position!" << endl;
            return;
        }
        if (pos == 1) {
            addAtFirstNode(name, price, available);
            return;
        }
        if (pos == size + 1) {
            addAtLastNode(name, price, available);
            return;
        }
        Node* newNode = new Node(name, price, available);
        Node* cur = head;
        for (int i = 1; i < pos - 1; i++) {
            cur = cur->next;
        }
        newNode->next = cur->next;
        newNode->prev = cur;
        cur->next->prev = newNode;
        cur->next = newNode;
        size++;
    }

    // Function to add a node at the end of the list
    void addAtLastNode(const string& name, int price, int available) {
        if (size == 0) {
            createCircularDL(name, price, available);
            return;
        }
        Node* newNode = new Node(name, price, available);
        newNode->next = head;
        newNode->prev = tail;
        tail->next = newNode;
        head->prev = newNode;
        tail = newNode;
        size++;
    }

    // Function to update the first node in the list
    void updateFirstNode(const string& name, int price, int available) {
        if (size == 0) {
            cout << "List is empty!" << endl;
            return;
        }
        head->name = name;
        head->price = price;
        head->available = available;
    }

    // Function to update the nth node in the list
    void updateNthNode(const string& name, int price, int available, int pos) {
        if (pos < 1 || pos > size) {
            cout << "Invalid position!" << endl;
            return;
        }
        Node* cur = head;
        for (int i = 1; i < pos; i++) {
            cur = cur->next;
        }
        cur->name = name;
        cur->price = price;
        cur->available = available;
    }

    // Function to update the last node in the list
    void updateLastNode(const string& name, int price, int available) {
        if (size == 0) {
            cout << "List is empty!" << endl;
            return;
        }
        tail->name = name;
        tail->price = price;
        tail->available = available;
    }

    // Function to remove the first node from the list
    void removeFirstNode() {
        if (size == 0) {
            cout << "List is empty!" << endl;
            return;
        }
        Node* temp = head;
        if (size == 1) {
            head = nullptr;
            tail = nullptr;
        }
        else {
            head = head->next;
            head->prev = tail;
            tail->next = head;
        }
        delete temp;
        size--;
    }

    // Function to remove the nth node from the list
    void removeNthNode(int pos) {
        if (pos < 1 || pos > size) {
            cout << "Invalid position!" << endl;
            return;
        }
        if (pos == 1) {
            removeFirstNode();
            return;
        }
        if (pos == size) {
            removeLastNode();
            return;
        }
        Node* cur = head;
        for (int i = 1; i < pos; i++) {
            cur = cur->next;
        }
        cur->prev->next = cur->next;
        cur->next->prev = cur->prev;
        delete cur;
        size--;
    }

    // Function to remove the last node from the list
    void removeLastNode() {
        if (size == 0) {
            cout << "List is empty!" << endl;
            return;
        }
        Node* temp = tail;
        if (size == 1) {
            head = nullptr;
            tail = nullptr;
        }
        else {
            tail = tail->prev;
            tail->next = head;
            head->prev = tail;
        }
        delete temp;
        size--;
    }

    // Function to print the circular doubly linked list
    void printCircularDL() const {
        if (size == 0) {
            cout << "List is empty!" << endl;
            return;
        }
        Node* cur = head;
        do {
            cout << "Name: " << cur->name << ", Price: " << cur->price << ", Available: " << cur->available << endl;
            cur = cur->next;
        } while (cur != head);
    }

    // Function to count the number of nodes in the list
    int countNodes() const {
        return size;
    }
};

int main() {
    LinkedList list;

    // Create the circular doubly linked list with the first node
    list.createCircularDL("Laptop", 20000000, 20);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Add a node at the beginning of the list
    list.addAtFirstNode("Handphone", 5000000, 50);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Add nodes at the end of the list
    list.addAtLastNode("Mouse", 1200000, 100);
    list.addAtLastNode("Keyboard", 2250000, 100);
    list.addAtLastNode("Headphone", 1800000, 100);
    list.addAtLastNode("Monitor", 4200000, 100);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Remove the first node from the list
    list.removeFirstNode();
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Remove the last node from the list
    list.removeLastNode();
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Add a node at the nth position in the list
    list.addAtNthNode("RTX 5090", 40000000, 15, 3);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Remove the nth node from the list
    list.removeNthNode(4);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Update the first node in the list
    list.updateFirstNode("Laptop", 19000000, 10);
    // Update the nth node in the list
    list.updateNthNode("Mouse", 1000000, 3, 2);
    // Update the last node in the list
    list.updateLastNode("Headphone", 1900000, 5);
    list.printCircularDL();
    cout << "==========\n\n\n";

    // Print the total number of nodes in the list
    cout << "Total nodes: " << list.countNodes() << endl;

    return 0;
}

File: 1 - LinkedList/Javascript/4-CircularDoubleLinkedList/circularDoubleLinkedList.js
/**
 * Class representing a node in the circular doubly linked list.
 */
class Node {
    /**
     * Create a node.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    constructor(name, price, available) {
        this.name = name;
        this.price = price;
        this.available = available;
        this.prev = null; // Pointer to the previous node
        this.next = null; // Pointer to the next node
    }
}

/**
 * Class representing a circular doubly linked list.
 */
class LinkedList {
    /**
     * Create a linked list.
     */
    constructor() {
        this.head = null; // Pointer to the head of the list
        this.tail = null; // Pointer to the tail of the list
        this.size = 0; // Size of the list
    }

    /**
     * Create a circular doubly linked list with the first node.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    createCircularDL(name, price, available) {
        const newNode = new Node(name, price, available);
        newNode.next = newNode;
        newNode.prev = newNode;
        this.head = newNode;
        this.tail = newNode;
        this.size = 1;
    }

    /**
     * Add a node at the beginning of the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    addAtFirstNode(name, price, available) {
        if (this.size === 0) {
            this.createCircularDL(name, price, available);
            return;
        }
        const newNode = new Node(name, price, available);
        newNode.next = this.head;
        newNode.prev = this.tail;
        this.head.prev = newNode;
        this.tail.next = newNode;
        this.head = newNode;
        this.size++;
    }

    /**
     * Add a node at the nth position in the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     * @param {number} pos - The position to insert the node.
     */
    addAtNthNode(name, price, available, pos) {
        if (pos < 1 || pos > this.size + 1) {
            console.log("Invalid position!");
            return;
        }
        if (pos === 1) {
            this.addAtFirstNode(name, price, available);
            return;
        }
        if (pos === this.size + 1) {
            this.addAtLastNode(name, price, available);
            return;
        }
        const newNode = new Node(name, price, available);
        let current = this.head;
        for (let i = 1; i < pos - 1; i++) {
            current = current.next;
        }
        newNode.next = current.next;
        newNode.prev = current;
        current.next.prev = newNode;
        current.next = newNode;
        this.size++;
    }

    /**
     * Add a node at the end of the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    addAtLastNode(name, price, available) {
        if (this.size === 0) {
            this.createCircularDL(name, price, available);
            return;
        }
        const newNode = new Node(name, price, available);
        newNode.next = this.head;
        newNode.prev = this.tail;
        this.tail.next = newNode;
        this.head.prev = newNode;
        this.tail = newNode;
        this.size++;
    }

    /**
     * Update the first node in the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    updateFirstNode(name, price, available) {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        this.head.name = name;
        this.head.price = price;
        this.head.available = available;
    }

    /**
     * Update the nth node in the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     * @param {number} pos - The position of the node to update.
     */
    updateNthNode(name, price, available, pos) {
        if (pos < 1 || pos > this.size) {
            console.log("Invalid position!");
            return;
        }
        let current = this.head;
        for (let i = 1; i < pos; i++) {
            current = current.next;
        }
        current.name = name;
        current.price = price;
        current.available = available;
    }

    /**
     * Update the last node in the list.
     * @param {string} name - The name of the item.
     * @param {number} price - The price of the item.
     * @param {number} available - The availability of the item.
     */
    updateLastNode(name, price, available) {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        this.tail.name = name;
        this.tail.price = price;
        this.tail.available = available;
    }

    /**
     * Remove the first node from the list.
     */
    removeFirstNode() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        const temp = this.head;
        if (this.size === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
            this.head.prev = this.tail;
            this.tail.next = this.head;
        }
        this.size--;
    }

    /**
     * Remove the nth node from the list.
     * @param {number} pos - The position of the node to remove.
     */
    removeNthNode(pos) {
        if (pos < 1 || pos > this.size) {
            console.log("Invalid position!");
            return;
        }
        if (pos === 1) {
            this.removeFirstNode();
            return;
        }
        if (pos === this.size) {
            this.removeLastNode();
            return;
        }
        let current = this.head;
        for (let i = 1; i < pos; i++) {
            current = current.next;
        }
        current.prev.next = current.next;
        current.next.prev = current.prev;
        this.size--;
    }

    /**
     * Remove the last node from the list.
     */
    removeLastNode() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        const temp = this.tail;
        if (this.size === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.next = this.head;
            this.head.prev = this.tail;
        }
        this.size--;
    }

    /**
     * Print all nodes in the circular doubly linked list.
     */
    printCircularDL() {
        if (this.size === 0) {
            console.log("List is empty!");
            return;
        }
        let current = this.head;
        do {
            console.log(`Name\t\t: ${current.name}\nPrice\t\t: ${current.price}\nAvailable\t: ${current.available}\n`);
            current = current.next;
        } while (current !== this.head);
    }

    /**
     * Utility function to count the number of nodes in the list.
     * @returns {number} The total number of nodes.
     */
    countNodes() {
        return this.size;
    }
}

// Example usage
const list = new LinkedList();
list.createCircularDL("Laptop", 20000000, 20);
list.printCircularDL();
console.log("==========\n");

list.addAtFirstNode("Handphone", 5000000, 50);
list.printCircularDL();
console.log("==========\n");

list.addAtLastNode("Mouse", 1200000, 100);
list.addAtLastNode("Keyboard", 2250000, 100);
list.addAtLastNode("Headphone", 1800000, 100);
list.addAtLastNode("Monitor", 4200000, 100);
list.printCircularDL();
console.log("==========\n");

list.removeFirstNode();
list.printCircularDL();
console.log("==========\n");

list.removeLastNode();
list.printCircularDL();
console.log("==========\n");

list.addAtNthNode("RTX 5090", 40000000, 15, 3);
list.printCircularDL();
console.log("==========\n");

list.removeNthNode(4);
list.printCircularDL();
console.log("==========\n");

list.updateFirstNode("Laptop", 19000000, 10);
list.updateNthNode("Mouse", 1000000, 3, 2);
list.updateLastNode("Headphone", 1900000, 5);
list.printCircularDL();
console.log("==========\n");

console.log(`Total nodes: ${list.countNodes()}`);